crates/ide_assists/src/handlers/unnecessary_async.rs

```diff
@@ -0,0 +1,220 @@
+use ide_db::{
+    assists::{AssistId, AssistKind},
+    base_db::FileId,
+    defs::{Definition, NameClass},
+    search::FileReference,
+};
+use syntax::{
+    ast::{self, HasName},
+    AstNode, SyntaxNode, TextSize,
+};
+
+use crate::{AssistContext, Assists};
+
+// Assist: unnecessary_async
+//
+// Removes the `async` mark from functions which have no `.await` in their body.
+// Looks for calls to the functions and removes the `.await` on the call site.
+//
+// ```
+// pub async f$0n foo() {}
+// pub async fn bar() { foo().await }
+// ```
+// ->
+// ```
+// pub  fn foo() {}
+// pub async fn bar() { foo() }
+// ```
+pub(crate) fn unnecessary_async(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let function: ast::Fn = ctx.find_node_at_offset()?;
+
+    // Do nothing if the cursor is not on the prototype. This is so that the check does not pollute
+    // when the user asks us for assists when in the middle of the function body.
+    // We consider the prototype to be anything that is before the body of the function.
+    let cursor_position = ctx.offset();
+    if cursor_position >= function.body().unwrap().syntax().text_range().start() {
+        return None;
+    }
+    // Do nothing if the function isn't async.
+    if let None = function.async_token() {
+        return None;
+    }
+    // Do nothing if the function has an `await` expression in its body.
+    if function.body()?.syntax().descendants().find_map(ast::AwaitExpr::cast).is_some() {
+        return None;
+    }
+
+    let async_range = function.async_token().unwrap().text_range();
+    let fn_name_pos = function.name()?.ident_token()?.text_range().start();
+
+    // Otherwise, we may remove the `async` keyword.
+    acc.add(
+        AssistId("unnecessary_async", AssistKind::QuickFix),
+        "Remove unnecessary async",
+        async_range,
+        |edit| {
+            // Remove async on the function definition.
+            edit.replace(async_range, "");
+
+            // Remove all `.await`s from calls to the function we remove `async` from.
+            for await_expr in
+                find_definitions(&ctx, ctx.sema.parse(ctx.file_id()).syntax(), fn_name_pos)
+                    .flat_map(|definition| find_all_references(ctx, &definition))
+                    // Keep only references that correspond to await expressions
+                    .filter_map(|(file_id, reference)| {
+                        find_await_expression(ctx, file_id, reference.range.start())
+                    })
+            {
```

Comment:
You seem to have misunderstood the intention of `find_definitions`, that function is meant to pick up everything defined at the cursor position.
You already have you definition here though, the function. So what you want to do here is take the function ast node and get a `hir` version from it, which can be done via `ctx.sema.to_def(&function)`, for that definition you then want to search for usages.

Reply:
The function looked very mysterious to me. I didn't know there was such a simple way of doing it.

Would my use of `find_nodes_at_offset_with_descend` be correct for `find_await_expression`? I would assume, from your comment that it is (also from the fact there is no conversation on it).

Thanks for your review!

Reply:
> Would my use of find_nodes_at_offset_with_descend be correct for find_await_expression? I would assume, from your comment, that it is (also from the fact there is no conversation on it). 

Ah I missed this earlier actually.

Technically it gets the job done yes, but there is a better more correct way(your current approach might fetch the wrong await expression in cases where they are nested), albeit a bit more work.

The usage search gives us back a bunch of file references which contain `ast::NameLike` (syntax nodes that are either `Name`, `NameRef` or `Lifetime`), we are only interested in the `NameRef`s here because we are looking for calls to our function. 

So for finding await expressions on our function calls what we can do here is take the `NameRef`, use this function here https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide_db/src/syntax_helpers/node_ext.rs#L18 to get the `ast::Path` it is part if, then do `.syntax().parent().and_then(ast::AwaitExpr::cast)` to get the corresponding await expression.

To find all method calls we can take the `NameRef` we have and just go with `.syntax().parent().and_then(ast::MethodCallExpr::cast)` and then again do the `.syntax().parent().and_then(ast::AwaitExpr::cast)`.
(Writing these steps outproperly as they require knowledge of the syntax tree structure of things, you can look the syntax trees of file by running the `Show Syntax Tree` command in r-a).

With that we have the await expressions, but these might be inside of macro expansions so a last step is to run `sema.original_ast_node(await_expr)` and take the result of  that. Now you finally have the await expression for which you can take the text ranges from.

Reply:
> your current approach might fetch the wrong await expression in cases where they are nested
I had included a couple tests for this specific case, since I was unsure. Though it was a simple test, it passed. I couldn't think of a way to have a different "kind" of nesting.

I had completely forgotten to search for the method calls. I'll add them and unit tests.

Thanks for all the pointers! I might not have much time to try my hands at it until next Tuesday, but I'll definitely tackle it!

Reply:
```rs
async fn foo() {}
async$0 fn bar() {}

async fn baz() {
   bar(foo().await).await()
}
```
I would've expected this test to fail with your current approach, but I just came to realize that we start the traversal from the `NameRef` token position, so that would actually still work out fine.

No worries, take your time with it 👍

Reply:
Hi,
I have modified as requested the `find_await_expression` to remove the `find_nodes_at_offset_with_descend`.
I still have questions, which I'll highlight with conversations, so they are close to the code they refer to.

I have added a test for macro calls, but it does not work. From what you wrote about macro expansions, I thought it would pass. I failed to see where it goes wrong, or maybe I am misunderstanding what `original_ast_node` does.